### DISCUSSION TITLE
Fold normalization into `laplacian_matrix`?

### DISCUSSION THREAD

**Discussion by rossbar** [networkx/networkx#5193] [Ideas]
Currently, the networkx.linalg package defines two functions: `nx.laplacian_matrix` and `nx.normalized_laplacian_matrix`. Looking closely at their implementations, `nx.normalized_laplacian_matrix` basically repeats the code in `nx.laplacian_matrix` and adds an extra normalization step before returning.

This made me think of an alternative organization for this functionality: rather than have two separate functions, the normalization could be folded into `nx.laplacian_matrix` and toggled with a new kwarg, e.g. `normalized: bool`. The advantages of this refactor would be:
 1. Improving code reuse (right now the laplacian computation is implemented independently in the two functions)
 2. Reduce the API footprint by removing a function name

I would also argue that this would actually make the normalization feature more discoverable, as I would expect users are more likely to look for Laplacian functionality via e.g. `nx.lapl<tab>` than to think to check `nx.normalized_<tab>`.

The refactor I'm proposing would involve deprecating the `normalized_laplacian_matrix` and adding a `normalize=` kwarg to `laplacian_matrix` instead. Thoughts?

**Comment by MridulS** [comment]
I'm -1 on deprecating `normalized_laplacian_matrix` but +1 on code reuse.

Folks looking for `normalized_laplacian_matrix` are usually *specifically* looking for the term Normalised Laplacian Matrix (at least from what I remember from my graph theory lectures). So I don't think it would make sense in this case to cut on the API footprint :)

But yeah if the code is being repeated which should definitely make that cleaner!

**Reply by rossbar** [reply]
> Folks looking for normalized_laplacian_matrix are usually specifically looking for the term Normalised Laplacian Matrix (at least from what I remember from my graph theory lectures). So I don't think it would make sense in this case to cut on the API footprint :)

I wasn't aware that this was specific terminology from the graph/network domain! In that case I agree - deprecation is not the way to go. Let's instead improve the code re-use as noted.
